### PR TITLE
trivial: dell: make failures to add GUIDs non-fatal (Fixes: #1396)

### PR DESCRIPTION
**Pull request by superm1**
Avoids situations that the TSS doesn't handle well such as TPM1.2.

Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/hughsie/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/dell/fu-plugin-dell.c [hunk 1]
```diff
@@ -749,8 +750,8 @@ fu_plugin_dell_detect_tpm (FuPlugin *plugin, GError **error)
 					    "Updating disabled due to TPM ownership");
 	}
 	/* build GUIDs from TSS strings */
-	if (!fu_plugin_dell_add_tpm_model (dev, error))
-		return FALSE;
+	if (!fu_plugin_dell_add_tpm_model (dev, &error_tss))
+		g_debug ("Couldn't build instances: %s", error_tss->message);
```
Comment (hughsie) [inline plugins/dell/fu-plugin-dell.c]: lower case, e.g. `could not build instance IDs` pls

Comment (superm1) [reply plugins/dell/fu-plugin-dell.c]: fixed.  Did this fix your problem then?

Comment (hughsie) [reply plugins/dell/fu-plugin-dell.c]: Gahh, no, still:

    # FuPluginDell-DEBUG: TPM Status: 0x101
    # FuPluginDell-DEBUG: Creating primary TPM GUID 0000-1.2 and secondary TPM GUID 0000-2.0
    # FuDevice-DEBUG: using 3ca929adb22a017654a24ce341b35205d2ec45cd for DELL-c8926cae-cee4-5d3b-80e5-70bd5ff1c101lu
    Bail out! FATAL-WARNING: Failed to create connection with service: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name is not activatable
    --- stderr ---
    ERROR:tcti:src/tss2-tcti/tcti-device.c:439:Tss2_Tcti_Device_Init() Failed to open device file /dev/tpm0: Permission denied 
    WARNING:tcti:src/tss2-tcti/tctildr.c:62:tcti_from_init() TCTI init for function 0x7effb4a3ffc0 failed with a000a 
    WARNING:tcti:src/tss2-tcti/tctildr.c:92:tcti_from_info() Could not initialize TCTI named: tcti-device 
    ERROR:tcti:src/tss2-tcti/tctildr-dl.c:150:tcti_from_file() Could not initialize TCTI file: libtss2-tcti-default.so 

    ** (/home/hughsie/Code/fwupd-stable/build/plugins/dell/dell-self-test:109069): WARNING **: 19:30:57.048: Failed to create connection with service: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name is not activatable

Comment (superm1) [reply plugins/dell/fu-plugin-dell.c]: Try as root. If it fails only as non root I think something else is going on.

Comment (hughsie) [reply plugins/dell/fu-plugin-dell.c]: Ohh, it works indeed as root.

Comment (superm1) [reply plugins/dell/fu-plugin-dell.c]: I think it might be something with tpm2 abrmd then not spinning up from dbus activation

Comment (superm1) [reply plugins/dell/fu-plugin-dell.c]: Do you see similar failures with `uefi-self-test` as non-root? I would think so.

**Comment by superm1** [comment]
^ see if that helps you

**Comment by hughsie** [comment]
Better, but still a failure. The UEFI plugin works :(

# FuDevice-DEBUG: using 3ca929adb22a017654a24ce341b35205d2ec45cd for DELL-c8926cae-cee4-5d3b-80e5-70bd5ff1c101lu
Bail out! FATAL-WARNING: Failed to create connection with service: GDBus.Error:org.freedesktop.DBus.Error.NoReply: Remote peer disconnected

** (./plugins/dell/dell-self-test:138484): WARNING **: 20:28:38.969: Failed to create connection with service: GDBus.Error:org.freedesktop.DBus.Error.NoReply: Remote peer disconnected
Trace/breakpoint trap (core dumped)

**Comment by superm1** [comment]
Oh I think I see why you're not hitting it in uefi.  `fu_uefi_pcrs_setup` doesn't actually run the 2.0 codepath if you have `.../sys/tpm0/pcrs` which is what happens in the 1.2 path.

This should break uefi for you too ^

**Comment by hughsie** [comment]
> This should break uefi for you too ^

It does indeed, now two failures :/

**Comment by superm1** [comment]
@diabonas can you comment on this failure scenario that it's failing to find a service? I would guess it's tpm2 abrmd. Should we just restrict that tcti in self tests to avoid it?

Hughsie can you share the whole self test log with new failures?

**Comment by hughsie** [comment]
[log.txt](https://github.com/fwupd/fwupd/files/3651810/log.txt)

**Comment by diabonas** [comment]
@superm1, @hughsie: Does this error also happen when running the self-test as root? Because the previous error
> ```
> ERROR:tcti:src/tss2-tcti/tcti-device.c:439:Tss2_Tcti_Device_Init() Failed to open device file > /dev/tpm0: Permission denied 
> WARNING:tcti:src/tss2-tcti/tctildr.c:62:tcti_from_init() TCTI init for function 0x7effb4a3ffc0 failed with a000a 
> WARNING:tcti:src/tss2-tcti/tctildr.c:92:tcti_from_info() Could not initialize TCTI named: tcti-device
> ERROR:tcti:src/tss2-tcti/tctildr-dl.c:150:tcti_from_file() Could not initialize TCTI file: libtss2-tcti-default.so
> ```
you got means that tpm2-tss is not packaged/installed correctly: `libtss2-tcti-default.so` is [meant to be a symlink to the TCTI you want to use](https://github.com/tpm2-software/tpm2-tss/issues/1482#issuecomment-516363966), so if tpm2-abrmd is installed, it should be a symlink to `libtss2-tcti-tabrmd.so`. The default installation [symlinks `libtss2-tcti-default.so` to `libtss2-tcti-device.so`](https://github.com/tpm2-software/tpm2-tss/pull/1486), so tpm2-abrmd is not used even if it is installed. If you're building tpm2-tss yourself from source, I would recommend simply deleting `libtss2-tcti-default.so`, in that case tpm2-tss will fall back to an auto-detection mechanism that does the right thing (use tpm2-abrmd if available, falling back to direct device access or the TPM simulator if unavailable).

It would be possible to force the self-tests to use the TPM simulator by setting the second argument of `Esys_Initialize`. With tpm2-tss 2.3, this is as easy as linking against `tss2-tctildr` and using
```
#include <tss2/tss2_tctildr.h>
// ...
TSS2_TCTI_CONTEXT *tcti_ctx;
Tss2_TctiLdr_Initialize ("mssim", &tcti_ctx);
Esys_Initialize (&ctx, tcti_ctx, NULL);
// ...
Esys_Finalize (&ctx);
Tss2_TctiLdr_Finalize(&tcti_ctx);
```
Note that this means that the TPM simulator must always be running during self-tests, a physical TPM will not be used at all any more even if available.

**Comment by hughsie** [comment]
> If you're building tpm2-tss yourself from source

This is with a unmodified F31 install.

**Comment by diabonas** [comment]
Fedora 31 packages `libtss2-tcti-default.so` in [tpm2-tss-devel](https://src.fedoraproject.org/rpms/tpm2-tss/blob/f65620047f9ebec79ca2089f10e4a2274b68900d/f/tpm2-tss.spec#_74), so if that package is installed, tpm2-abrmd will never work by default on this distribution (regardless of fwupd). This is a packaging bug that should be [reported against Fedora](https://bugzilla.redhat.com/enter_bug.cgi?product=Fedora&version=rawhide&component=tpm2-tss). In the meantime, workarounds are either to temporarily remove tpm2-tss-devel during testing or to delete `/usr/lib/libtss2-tcti-default.so`, which is harmless but quick-and-dirty since it messes around with files installed by the package manager.

On the other hand using `libtss2-tcti-default.so` correctly seems to be a bit of a pitfall, so we should revisit the upstream decision in tpm2-tss to install it by default. I have opened https://github.com/tpm2-software/tpm2-tss/issues/1525 there to track this issue.

**Comment by superm1** [comment]
I believe hughsie mentioned above that running as root it's actually okay. It's only a failure with that warning that causes self test failure as a user.

Since this seems to be a packaging problem, I'm inclined to just add a check whether running as root in the self tests and paper over this. The CI runs as root anyway so we'll keep exercising it.

**Comment by diabonas** [comment]
> Since this seems to be a packaging problem, I'm inclined to just add a check whether running as root in the self tests and paper over this. The CI runs as root anyway so we'll keep exercising it.

Fine with me, this seems to be the only way to guarantee that the TPM is accessible anyway since we don't know whether tpm2-abrmd is installed or not. On the other hand, self-tests should probably be run against the TPM simulator anyway by default, so forcing the simulator TCTI seems feasible as well. This is what I originally had in mind, but as indicated in https://github.com/fwupd/fwupd/pull/1311#discussion_r317702157, starting the simulator from within the self-test is hard, so I abandoned the idea.

**Comment by superm1** [comment]
> > Since this seems to be a packaging problem, I'm inclined to just add a check whether running as root in the self tests and paper over this. The CI runs as root anyway so we'll keep exercising it.
> 
> Fine with me, this seems to be the only way to guarantee that the TPM is accessible anyway since we don't know whether tpm2-abrmd is installed or not. On the other hand, self-tests should probably be run against the TPM simulator anyway by default, so forcing the simulator TCTI seems feasible as well. This is what I originally had in mind, but as indicated in [#1311 (comment)](https://github.com/fwupd/fwupd/pull/1311#discussion_r317702157), starting the simulator from within the self-test is hard, so I abandoned the idea.

In a CI context you have a pretty good control of the environment.  It's nice that it's packaged and easy to start up from a CI context at least in Arch.

I guess as long as we still have CI running against the simulator on at least one target that's the important part.

**Comment by superm1** [comment]
@hughsie can you try now?  I think this should adequately solve it for your local use case while still letting us run the test in CI.

**Comment by hughsie** [comment]
Works for me, thanks!

**Comment by superm1** [comment]
OK, just rebased to clean up commit messages, i'll merge once CI passes.

**Pull request merged**